### DISCUSSION TITLE
How to embed a copy-to-clipboard button?

### DISCUSSION THREAD

**Discussion by solzard** [zauberzeug/nicegui#250] [Q&A] [answered]
Is it possible to create a copy-to-clipboard button such as one generated on the right top of code blocks written with GitHub-flavored markdown?
![gfm-example](https://user-images.githubusercontent.com/29563481/212876766-c3bc166a-1e90-4c91-bc58-647d913eb6f6.jpg)

**Comment by falkoschindler** [comment, marked as answer]
In JavaScript you can write to the clipboard as follows:

```js
navigator.clipboard.writeText("Some text");
```

So you can easily integrate it into a NiceGUI app:

```python3
async def clip():
    await ui.run_javascript('navigator.clipboard.writeText("Some text")')

ui.button('Clip', on_click=clip)
```

**Reply by solzard** [reply]
I'll leave a snippet that accepts an arbitrary text and pop up a notification.
The idea below to use `functools.partial` is from [Async event handlers - NiceGUI Reference](https://nicegui.io/reference#async_event_handlers).

> Note: You can also pass a functools.partial into the on_click property to wrap async functions with parameters.

```python
#!/usr/bin/env python3
from functools import partial

from nicegui import ui


def add_clip_button(title: str, text: str) -> None:
    ui.button(f"Clip {title}", on_click=partial(clip, text))


async def clip(text: str) -> None:
    await ui.run_javascript(
        f'navigator.clipboard.writeText("{text}")',
        respond=False,
    )
    ui.notify(
        message="Copied to clipboard",
        position="top-right",
        color="green",
    )


title_to_text = {
    "GitHub": "https://github.com",
    "Python": "https://python.org",
    "NiceGUI": "https://nicegui.io",
}

for title, text in title_to_text.items():
    add_clip_button(title, text)

ui.run()
```

**Comment by DarakDuVal** [comment]
Update: received an error awaiting the Javascript to complete and using respond key:
```py
def clip(text: str) -> None:
    ui.run_javascript(f'navigator.clipboard.writeText("{text}")')
    ui.notify(message='Copied to clipboard', color='green')
```

**Reply by falkoschindler** [reply]
Yes, the API for [`ui.run_javascript`](https://nicegui.io/documentation/run_javascript) has changed since last year.

**Reply by DarakDuVal** [reply]
Yes, but thanks to great package maintainers, the debug message was very clear and easy to follow. Thanks for that @falkoschindler !

**Reply by frankvp11** [reply]
were you able to figure it out? Conveniently I ran into a similar issue, and it doesn't appear to be showing up in my clipboard?

**Reply by falkoschindler** [reply]
@frankvp11 Maybe it's related to [this limitation in Safari](https://github.com/zauberzeug/nicegui/discussions/2364)?
I recommend implementing a copy-to-clipboard button using the new `js_handler` like we do for the `ui.code` element:
https://github.com/zauberzeug/nicegui/blob/b7b30b2c801c6c2fe46efb0af6eef34205884163/nicegui/elements/code.py#L30-L32

**Comment by lij55** [comment]
![Screenshot From 2025-02-18 18-35-31](https://github.com/user-attachments/assets/810b758e-438c-4da4-85a2-fbcea95a6f16)
I got this error message with funcion
```py
def clip():
    result = output.value
    ui.run_javascript('navigator.clipboard.writeText("{}")'.format(result))
ui.button('Clip', on_click=clip)
```
is it possible to config this in python instead of js?

updated:
I have tried and found if there is a newline(`\n`) in the string, it will raise this error. it works well without `\n`

**Reply by falkoschindler** [reply]
Hi @lij55,

Your code works in principle. But depending on the content of `output.value` (e.g. containing a double quote), the resulting JavaScript code might be invalid.
A better way is to use the json module to convert `result` to a JavaScript representation:
```py
def clip():
    result = '"foo"'
    ui.run_javascript('navigator.clipboard.writeText({})'.format(json.dumps(result)))

ui.button('Clip', on_click=clip)
```

Note that since version 1.4.20 there's a `ui.clipboard` feature to simplify your code quite a bit:
```py
def clip():
    result = '"bar"'
    ui.clipboard.write(result)

ui.button('Clip', on_click=clip)
```

**Reply by lij55** [reply]
fantastic!! `ui.clipboard.write` works!!
